### PR TITLE
simplify publish workflow with explicit matrix

### PR DESCRIPTION
**Pull request by deitch**
Simplifies the `publish.yml` workflow by explicitly using the matrix for arch, and removing a lot of the redundant `if` conditions.

I would like to do the same to `build.yml`, but it is harder there. Because those _build_ packages but do not _publish_ them , the packages are only available on the linuxkit cache on the specific runner. We would have to get the cached packages off of the runner and onto another runner before we could start doing `make eve`. The easiest thing to do is what we do now: run `make eve` as part of the same job on the same runner.

If we could use `lkt pkg build` with its docker context support for remote runners, then it already would centralize it all to a single cache. But we do not do that here; it would require setting up docker contexts on machines that are not necessarily remotely accessible from the "main" runner.

This is a step forward for now.

### CODE REVIEW

**Comment by deitch** [comment]
Update: I figured out how to do it in the `build.yml` workflow with GHA cache. I usually use upload/download artifacts, so I might have gotten this incorrect. We will let it run and see.

File: .github/workflows/build.yml
Comment (giggsoff) [inline .github/workflows/build.yml:68]: Should it be kvm instead of eve?

File: .github/workflows/build.yml
Comment (giggsoff) [inline .github/workflows/build.yml:80]: As I understand you should define VERSION variable. Also I can see TAG variable few lines below.

File: .github/workflows/build.yml
Comment (giggsoff) [inline .github/workflows/build.yml:73]: In that case you will load linuxkit cache, which is good. But we still need some packages to be available in docker, as we discussed previously https://github.com/lf-edge/eve/pull/2727#issuecomment-1236695661. How to get them in place here?

File: .github/workflows/build.yml
Comment (deitch) [reply .github/workflows/build.yml:68]: Shoot, yes, nice catch. Way too much copy-paste. I am fixing.

File: .github/workflows/build.yml
Comment (deitch) [reply .github/workflows/build.yml:73]: That is a very good point. I hadn't thought about that.

How does it work in `publish.yaml`, where it currently uses the remote runners to build and publish packages `make pkgs`, and then the local runner to `make eve`. Is it relying on the fact that it already had run `make pkgs` there?

Comment (giggsoff) [reply .github/workflows/build.yml:69]: Correct, we did make pkgs before, so have all in docker. Also we publish packages firstly, so we can pull them from DockerHub.

File: .github/workflows/build.yml
Comment (deitch) [reply .github/workflows/build.yml]: Fixed

File: .github/workflows/build.yml
Comment (giggsoff) [reply .github/workflows/build.yml:69]: Probably [cache-export](https://github.com/lf-edge/eve/blob/e6c69cc972828a70e6e788c4cea02b4a40ce4f53/Makefile#L614-L616) may help us to extract images from linuxkit cache.

Comment (deitch) [reply .github/workflows/build.yml:69]: We maintain the list of required docker packages in the `Makefile`:

```makefile
PKGS_DOCKER_LOAD=mkconf mkimage-iso-efi mkimage-raw-efi mkrootfs-ext4 mkrootfs-squash
```

We could do a `make load-docker-packages` which would export/import. I am worried it is a bit convoluted, but we can try it. What do you think, @giggsoff ?

File: .github/workflows/build.yml
Comment (giggsoff) [inline .github/workflows/build.yml:58]: I can see the wrong order of builds, seems we should add `needs: packages` for `eve` job.

**Comment by deitch** [comment]
I also see build failures on the arm64 matrix for `docker tag evebuild:danger`, mainly because it never loaded `make eve` into docker after the cache import/export. We could do the same thing.

Didn't we originally want to do that directly via `lkt cache export`? Why didn't we?

File: .github/workflows/build.yml
Comment (giggsoff) [reply .github/workflows/build.yml:69]: Agree, not so clear, but we can try at least.

File: .github/workflows/build.yml
Comment (deitch) [reply .github/workflows/build.yml:58]: Correct. Fixed.

**Comment by giggsoff** [comment]
> I also see build failures on the arm64 matrix for `docker tag evebuild:danger`, mainly because it never loaded `make eve` into docker after the cache import/export.

Not sure, I can see that it is loaded: https://github.com/lf-edge/eve/actions/runs/3105850281/jobs/5031980791#step:5:501
`Loaded image: lfedge/eve:0.0.0-pr-27dd8ef3-kvm-arm64`. But it use arch suffix because we decided to not load images for another arch into docker without suffix.

**Comment by deitch** [comment]
> Not sure, I can see that it is loaded: https://github.com/lf-edge/eve/actions/runs/3105850281/jobs/5031980791#step:5:501
Loaded image: lfedge/eve:0.0.0-pr-27dd8ef3-kvm-arm64. But it use arch suffix because we decided to not load images for another arch into docker without suffix.

Oh is that it? So I could have just changed the tag? Well, it doesn't matter. We just used it to `docker save` to a tar file anyways; if this `cache-export` works, then fine.

File: .github/workflows/build.yml
Comment (giggsoff) [inline .github/workflows/build.yml:95]: Will it work without setting of ZARCH to export arm64 image on amd64 host?

Comment (deitch) [reply .github/workflows/build.yml:95]: Haha! That _is_ the error! Going to push now.

**Comment by deitch** [comment]
The `unit tests` failure is a spurious one

**Comment by deitch** [comment]
Well, CI is very happy. I would like to download the artifacts to double-check them, but I am pretty pleased.

@giggsoff is it possible that the docker load doesn't matter, because the tools are local anyways? if it were an issue, CI would fail?

**Comment by giggsoff** [comment]
> Well, CI is very happy. I would like to download the artifacts to double-check them, but I am pretty pleased.

Something wrong: we do not expect [evebuild/danger:pr-xen](https://hub.docker.com/layers/evebuild/danger/pr-xen/images/sha256-c4b3a08d413102f27709def0118504963d7c81adf63a0c012fa2dc43efe9f194?context=explore) without pr number. Looks like we miss PR_ID variable somehow.

> @giggsoff is it possible that the docker load doesn't matter, because the tools are local anyways? if it were an issue, CI would fail?

We did not update tools, so they just loaded from docker hub. In that case CI will not fail.

**Comment by deitch** [comment]
> Something wrong: we do not expect [evebuild/danger:pr-xen](https://hub.docker.com/layers/evebuild/danger/pr-xen/images/sha256-c4b3a08d413102f27709def0118504963d7c81adf63a0c012fa2dc43efe9f194?context=explore) without pr number. Looks like we miss PR_ID variable somehow.

I will check

> We did not update tools, so they just loaded from docker hub. In that case CI will not fail

So to test it, we need a commit with a change to one of the tools, nothing major, just enough to make it needs pull local? OK, we can do that after the previous thing is fixed.

**Comment by deitch** [comment]
Downloaded the artifacts, expanded the tar files, all look right to me.

**Comment by deitch** [comment]
I updated to the latest linuxkit with support for setting cache location via env var, and then set it consistently in both jobs. Plan now:

1. Let CI run, be sure it is green
2. Check that the cache worked
3. Do a "no real changes" commit that forces new versions of some of the stuff that normally needs to load into docker; see that it is green
4. Remove that commit, wait for CI... we will be able to merge in

File: .github/workflows/eden.yml
Comment (giggsoff) [inline .github/workflows/eden.yml:85]: You did not set env for this workflow.

Comment (deitch) [reply .github/workflows/eden.yml:85]: We don't need it. Isn't eden using that weird `evedanger` repo to pull it all down?

Mind, if we used cache, we could get rid of `evedanger`! OK, I will add that here.

Comment (giggsoff) [reply .github/workflows/eden.yml:85]: I mean you change directory to cleanup, but did not set env to use the directory. I think it would be better to not touch this workflow now

Comment (deitch) [reply .github/workflows/eden.yml:85]: Well, it cannot hurt to use the cache. Let's at least add that.

Comment (deitch) [reply .github/workflows/eden.yml]: I changed my mind. You are right. The `eden.yml` flow is different. Let's do that in a different PR.

Comment (giggsoff) [reply .github/workflows/eden.yml]: Will it be more clean to use env value here instead of path?

Comment (deitch) [reply .github/workflows/eden.yml]: Eventually. For now, I removed all of the eden changes.

**Comment by deitch** [comment]
> Let CI run, be sure it is green

✅ 

> Check that the cache worked

✅ 

> Do a "no real changes" commit that forces new versions of some of the stuff that normally needs to load into docker; see that it is green

Coming next

**Comment by deitch** [comment]
Yes, now it is the docker image issue that we expected @giggsoff ; just need to load them up.

**Comment by deitch** [comment]
I added some make targets. I am not madly in love with the design, but it should work.

**Comment by deitch** [comment]
Found a cleaner, more make-native way to do it.

File: .github/workflows/build.yml
Comment (giggsoff) [inline .github/workflows/build.yml:75]: Is it safe to clean cache inside Clean step in case of using this step? Will we end with cache cleanup while we potentially still need it in another runners?

Comment (deitch) [reply .github/workflows/build.yml:75]: I don't see why not. Each job in a runner is completely independent. If you don't get rid of it, you will lose it anyways. Except on the persistent arm64 self-hosted runner, but that really should be ephemeral anyways.

Comment (deitch) [reply .github/workflows/build.yml:75]: For now, it looks good. If this all passes, we merge it in. If it creates problems in the future, we worry about it then.

Comment (giggsoff) [reply .github/workflows/build.yml:75]: I mean we have a step with `rm -rf ~/.linuxkit` in the end. But with new step you put here cache update will happen after cleaning (in post action). So my understanding here is if we will start one `packages` job for amd64-kvm but for some reason amd64-xen will delayed (for example we will hit limit of parallel runners), we will clean the cache before amd64-xen will be started. Correct me if I am wrong, as it depends on the internal logic of actions/cache.

Comment (deitch) [reply .github/workflows/build.yml:75]: Each of those jobs runs in their own runner. So rm -rf on one runner has no impact on the other. It's why I wanted to use the matrix and run them in parallel, rather than the serial we had before.

That is for the linuxkit cache. The GitHub cache stays until it expires it, based on time or size.

Comment (giggsoff) [reply .github/workflows/build.yml:75]: Got it, thanks for explaining!

File: .github/workflows/build.yml
Comment (rvs) [inline .github/workflows/build.yml:39]: I see you added zstd to the build VM image -- can you please put a comment here prompting to remove this block?

**Review by rvs — APPROVED** [review]
LGTM in general -- but please take care of some of the nits before merging

File: pkg/mkconf/Dockerfile
Comment (giggsoff) [reply pkg/mkconf/Dockerfile:14]: This is dummy update. But it helps us to check what will happen in case of real update. And we see the problem. In packages workflow we build packages required for tools on arm64 for host arch (again, arm64). But in composing job we need packages for another host arch (amd64) and they are not available.
I think we should build packages required for tools in runner (job) where we do composing.

Comment (deitch) [reply pkg/mkconf/Dockerfile:14]: Exactly. Once it all runs clean, I will remove that commit and push again.

File: .github/workflows/build.yml
Comment (deitch) [reply .github/workflows/build.yml:39]: I can. When do you think that build will apply, so we can remove it entirely?

**Comment by deitch** [comment]
Close, but it still doesn't 100% work. We now have the package builds on the actual arch (arm64 on arm64, amd64 on amd64) which is good, but eve builds on whatever runners are available, and in a matrix, so we get parallelism. Even better.

The runner loads into the 5 or so images we need from the restored linuxkit cache, but only if the arch for which we are building is the arch on which we are running. Having an arm64 mkconf is useless on amd64 runner, as we need to _run it_, not put it in a composed eve image.

But that means that the runner now has _no_ copy of those packages, so the build on arm64 fails.

Thinking best way to solve this. More coming.

**Comment by deitch** [comment]
Take 35. Let's see if this does it.

**Comment by deitch** [comment]
Huzzah! All works as expected. Going to remove dummy commit, and then this is ready to go.

**Comment by deitch** [comment]
Ready for reviews and merge.

File: .github/workflows/build.yml
Comment (giggsoff) [inline .github/workflows/build.yml:89]: It seems we will load cache twice (amd64 and arm64) for arm64, which allow us to have all packages in place. But should we load the same cache (amd64) twice for amd64?

**Comment by giggsoff** [comment]
I can see from last run, that step `update linuxkit cache for our arch` was skipped for arm64 build https://github.com/lf-edge/eve/actions/runs/3138378758/jobs/5097727246. Not sure that it is what we expect.

File: .github/workflows/build.yml
Comment (giggsoff) [reply .github/workflows/build.yml:89]: Should it be `if: ${{ matrix.arch != 'amd64' }}` or so?

Comment (deitch) [reply .github/workflows/build.yml]: Yes, you saw it the same time I did. I got the logic wrong, but it is fixed now.

File: .github/workflows/build.yml
Comment (giggsoff) [inline .github/workflows/build.yml:47]: Small question here: in case of need to re-run the workflow (for example because of connectivity issues), will we hit the problems with the old cache? Probably it would be nice to add `github.run_number` here?

Comment (deitch) [reply .github/workflows/build.yml:47]: I think that might create problems, rather than resolve them. I worked up the scenarios:

* fails during packages: no cache is created, so it will have a cache miss on next run and create it: ✅ 
* fails during eve, rerun all: packages will load the cache, which is good; eve rerun will use the same cache: ✅ 
* fails during eve, rerun just eve: will use the cached output: ✅ 

Either way, we have the desired behaviour. What happens if we add the run number?

* fails during packages: no cache is created, so it will have a cache miss on next run and create it: ✅ 
* fails during eve, rerun all: packages will have a cache miss (different run_number), so it will recreate all, a bit inefficient, but it works: inefficient ✅ 
* fails during eve, rerun eve: when packages ran, it had a different run number, so there is no cache, eve will fail ❌ 

Adding run number either makes it fail, or makes it a bit more inefficient. I cannot find a scenario where leaving it out fails. 🤷‍♂️

**Comment by deitch** [comment]
@rvs and @eriknordmark I think this is ready.

**Comment by eriknordmark** [comment]
This past eden tests and ztests. Were there some modifications there for testing that you need to remove before we pull this? And are all the nits taken care of?

**Comment by deitch** [comment]
> Were there some modifications there for testing that you need to remove before we pull this? And are all the nits taken care of?

All nits were taken care of @eriknordmark , the test commit passed, and I removed the test commit before you ran this. So this should be good to go.

**Pull request merged**
